File: typescript/src/signing.ts
/**
 * Mandate signing and verification, backed by Node's built-in Ed25519 (no
 * external crypto dependency). Keys are raw 32-byte values, wire-compatible with
 * the Python port: same canonical bytes, same RFC-8032 signatures, same base64
 * envelope.
 */

import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as nodeSign,
  verify as nodeVerify,
} from "node:crypto";
import { canonicalize } from "./canonical.ts";

// Fixed DER prefixes for a raw Ed25519 seed / public key.
const PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export interface SignedMandate {
  mandate: Record<string, unknown>;
  signature: { alg: "Ed25519"; public_key: string; value: string };
}

export function generateKeypair(): { privateKey: Buffer; publicKey: Buffer } {
  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  const seed = privateKey.export({ format: "der", type: "pkcs8" }).subarray(-32);
  const pub = publicKey.export({ format: "der", type: "spki" }).subarray(-32);
  return { privateKey: Buffer.from(seed), publicKey: Buffer.from(pub) };
}

function privFromSeed(seed: Buffer) {
  return createPrivateKey({
    key: Buffer.concat([PKCS8_PREFIX, seed]),
    format: "der",
    type: "pkcs8",
  });
}

function pubFromRaw(pub: Buffer) {
  return createPublicKey({
    key: Buffer.concat([SPKI_PREFIX, pub]),
    format: "der",
    type: "spki",
  });
}

export function publicKeyFromSeed(seed: Buffer): Buffer {
  const der = createPublicKey(privFromSeed(seed))
    .export({ format: "der", type: "spki" })
    .subarray(-32);
  return Buffer.from(der);
}

export function signMandate(
  mandate: Record<string, unknown>,
  seed: Buffer,
): SignedMandate {
  const publicKey = publicKeyFromSeed(seed);
  const signature = nodeSign(null, canonicalize(mandate), privFromSeed(seed));
  return {
    mandate,
    signature: {
      alg: "Ed25519",
      public_key: publicKey.toString("base64"),
      value: Buffer.from(signature).toString("base64"),
    },
  };
}

export function verifySignature(signed: SignedMandate): boolean {
  const sig = signed?.signature;
  if (!sig || sig.alg !== "Ed25519") return false;
  try {
    const pub = pubFromRaw(Buffer.from(sig.public_key, "base64"));
    return nodeVerify(
      null,
      canonicalize(signed.mandate),
      pub,
      Buffer.from(sig.value, "base64"),
    );
  } catch {
    return false;
  }
}

File: python/mandatekit/signing.py
"""
Mandate signing and signature verification.

The signing key is a 32-byte Ed25519 seed that stays on the caller's machine.
`sign_mandate` attaches a detached signature plus the public key so a verifier
needs nothing but the signed document itself.
"""

import base64
import os
from typing import Dict, Tuple

from . import ed25519
from .canonical import canonicalize


def generate_keypair() -> Tuple[bytes, bytes]:
    """Return (private_seed, public_key) as raw 32-byte values."""
    seed = os.urandom(32)
    return seed, ed25519.publickey(seed)


def _b64(b: bytes) -> str:
    return base64.b64encode(b).decode("ascii")


def _unb64(s: str) -> bytes:
    return base64.b64decode(s)


def sign_mandate(mandate: Dict, private_key: bytes) -> Dict:
    """Wrap a mandate dict in a signed envelope."""
    public_key = ed25519.publickey(private_key)
    signature = ed25519.sign(canonicalize(mandate), private_key, public_key)
    return {
        "mandate": mandate,
        "signature": {
            "alg": "Ed25519",
            "public_key": _b64(public_key),
            "value": _b64(signature),
        },
    }


def verify_signature(signed: Dict) -> bool:
    """True iff the envelope's signature matches its mandate. Never raises."""
    sig = signed.get("signature") or {}
    if sig.get("alg") != "Ed25519":
        return False
    try:
        public_key = _unb64(sig["public_key"])
        value = _unb64(sig["value"])
    except Exception:
        return False
    return ed25519.verify(value, canonicalize(signed["mandate"]), public_key)
